### PR TITLE
Add support for Northeastern University 🎉🎉🎉🎉🎉

### PR DESCRIPTION
**Pull request by ryanhugh**
YOOO!!

Some background:
I am a 4th year CS student at Northeastern University and created [Search NEU](https://searchneu.com) which is a search box for all the classes at NEU. Search NEU is written in Node.js and scrapes Northeastern's registration system (Banner, same as GW) to get all the data. This PR incorporates those scrapers into Semester.ly to add support for NEU to Semester.ly. The Search NEU scrapers are entirely written in Node.js and have no external dependencies outside of node_modules. 

Major notes:
- published Search NEU onto NPM (https://www.npmjs.com/package/searchneu)
- added searchneu to Semesterly's package.json
- Creating a nodejs subprocess inside the NEU scrapers to call a small JS script that calls require('searchneu'). This nodejs subprocess saves to schools/neu/data/courses.json

I'm going to change the communication between Search NEU and Semester.ly in the future so we can add support for more schools easily. Right now everything is just hardcoded for NEU Fall 2017. Also some other small stuff to cleanup. Let me know what you guys think!!

Example:
image

### CODE REVIEW

**Comment by mnm364** [comment]
Awesome stuff! Just real quick, if you want to merge move the TODOs into a comment rather than the merge message. Ill give this a look and ping you asap!

**Comment by ryanhugh** [comment]
TODO:

- [ ] Add support for NEU textbooks ( http://northeastern.bncollege.com/webapp/wcs/stores/servlet/TBWizardView?storeId=17555&catalogId=10001&langId=-1 ). Northeastern uses the Barnes and Noble website which hopefully isn't too hard to add support for. 
- [ ] Add support for more terms (dynamically scrape semesters?)
- [ ] other small stuff

**Comment by mnm364** [comment]
Is there anyway for you to narrow down the `course_code_regex` field in the `config.json`? We use that to provide links to courses within descriptions, requisites, etc.

**Comment by ryanhugh** [comment]
Done! Made sure that regex matches all the course codes at NEU. Also fixed the linting errors.

**Comment by ryanhugh** [comment]
Also here's some info about the Search NEU Cache (and the addition to .gitignore). When you run the Search NEU scrapers in development mode they save the http response for every request in a cache folder. Then, if you run the scrapers in development mode again, they will just load from these files instead of hitting the server. Loading from the files is a lot faster than hitting the server which helps a lot when developing/making tweaks to the scrapers. In production mode, the cache is skipped and all the requests will always be sent to the server. Right now it will run in development mode when running locally and production mode if process.env.CI or process.env.PROD are present. Overall it works really well, the only gotcha is if you are running it with cache enabled and are wondering why it isn't downloading the latest data from the server. Let me know if you want to change it up. I could totally disable the cache entirely when running as part of Semesterly.

**Comment by mnm364** [comment]
I like that a lot... like a lot! I want to add that as part of Semesterly's parsing development as well. Have been struggling with this issue for a while and this seems like an elegant and simple way to handle it! Thanks for the heads up as well!

**Comment by ryanhugh** [comment]
No problem - I'l leave it in!

All the code behind it is in these two files
https://github.com/ryanhugh/searchneu/blob/master/backend/scrapers/cache.js
https://github.com/ryanhugh/searchneu/blob/master/backend/scrapers/request.js

Happy to talk more about it if you are interested

File: parsing/schools/neu/config.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
+{
+  "school": {
+    "code": "neu",
+    "name": "Northeastern University"
+  },
+  "course_code_regex": "([A-Z]{0,4} [0-9]{4})",
```
Comment (mnm364) [inline parsing/schools/neu/config.json]: add`$` to end of regex match

File: parsing/schools/neu/courses.py [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+
+from __future__ import absolute_import, division, print_function
+
+import re
+import sys
+import os
+import subprocess
+
+from bs4 import NavigableString, Tag
```
Comment (mnm364) [inline parsing/schools/neu/courses.py]: Why do we import this?

**Comment by mnm364** [comment]
Running locally I get.

```
<--- JS stacktrace --->

==== JS stack trace =========================================

Security context: 0x884735cf781 <JS Object>
    1: standardizeInputConfig [/home/mike/semesterly/node_modules/searchneu/dist/backend/scrapers/request.js:~783] [pc=0x55d476b7a9] (this=0xca797080551 <a RequestInput with map 0x161541fdbdf1>,config=0x2b0430447e69 <an Object with map 0x2a741e907119>)
    2: _callee5$ [/home/mike/semesterly/node_modules/searchneu/dist/backend/scrapers/request.js:764] [pc=0x55d31bb98b] (this=0xca797080551 <a Req...

FATAL ERROR: CALL_AND_RETRY_LAST Allocation failed - JavaScript heap out of memory
 1: node::Abort() [node]
 2: 0x109f13c [node]
 3: v8::Utils::ReportApiFailure(char const*, char const*) [node]
 4: v8::internal::V8::FatalProcessOutOfMemory(char const*, bool) [node]
 5: v8::internal::Factory::NewFillerObject(int, bool, v8::internal::AllocationSpace) [node]
 6: v8::internal::Runtime_AllocateInTargetSpace(int, v8::internal::Object**, v8::internal::Isolate*) [node]
```

My computer is not necessarily weak, anyway we can make this use less memory?

**Comment by mnm364** [comment]
Also would it be possible to silence some of the output? Maybe pass a versbosity arg through command line arg over subprocess?

File: parsing/schools/neu/courses.py [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+
+from parsing.library.base_parser import BaseParser
+from parsing.library.exceptions import ParseError
+from parsing.library.utils import safe_cast, dict_filter_by_dict
+from semesterly.settings import get_secret
+
+
+class Parser(BaseParser):
+   
+    def __new__(cls, *args, **kwargs):
```
Comment (mnm364) [inline parsing/schools/neu/courses.py]: i know this is in the template but if you dont have any secrets to load, no need to have a `__new__` method

File: parsing/schools/neu/courses.py [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+    # Required to prevent the python code from writing courses.json.
+    # The JS script saves the courses.json when using search neu scrapers.
+    def end(self):
+      pass
+
+    def start(self,
+              years_and_terms_filter=None,
+              departments_filter=None,
+              verbosity=3,
+              textbooks=None):
```
Comment (mnm364) [inline parsing/schools/neu/courses.py]: Same here, I know this is in the template, but can just make this `**kwargs` (not a big deal tho)

File: parsing/schools/neu/courses.py [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+import sys
+import os
+import subprocess
+
+from bs4 import NavigableString, Tag
+
+from parsing.library.base_parser import BaseParser
+from parsing.library.exceptions import ParseError
+from parsing.library.utils import safe_cast, dict_filter_by_dict
+from semesterly.settings import get_secret
```
Comment (mnm364) [inline parsing/schools/neu/courses.py]: We arent using a secret for `neu`

File: parsing/schools/neu/courses.py [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+
+import re
+import sys
+import os
+import subprocess
+
+from bs4 import NavigableString, Tag
+
+from parsing.library.base_parser import BaseParser
+from parsing.library.exceptions import ParseError
```
Comment (mnm364) [inline parsing/schools/neu/courses.py]: not needed

File: parsing/schools/neu/courses.py [hunk 7]
```diff
@@ -0,0 +1,64 @@
...
+from parsing.library.utils import safe_cast, dict_filter_by_dict
```
Comment (mnm364) [inline parsing/schools/neu/courses.py]: not needed

File: parsing/schools/neu/config.json (hunk 1 above)
Comment (ryanhugh) [reply parsing/schools/neu/config.json]: Done

File: parsing/schools/neu/courses.py (hunk 2 above)
Comment (ryanhugh) [reply parsing/schools/neu/courses.py]: Removed

File: parsing/schools/neu/courses.py (hunk 6 above)
Comment (ryanhugh) [reply parsing/schools/neu/courses.py]: Removed

File: parsing/schools/neu/courses.py (hunk 7 above)
Comment (ryanhugh) [reply parsing/schools/neu/courses.py]: Removed

File: parsing/schools/neu/courses.py (hunk 5 above)
Comment (ryanhugh) [reply parsing/schools/neu/courses.py]: Removed

File: parsing/schools/neu/courses.py (hunk 3 above)
Comment (ryanhugh) [reply parsing/schools/neu/courses.py]: Removed

File: parsing/schools/neu/courses.py (hunk 4 above)
Comment (ryanhugh) [reply parsing/schools/neu/courses.py]: Changed

**Comment by mnm364** [comment]
@ryanhugh  I think you deleted all of Hopkins Evals by accident.

**Comment by ryanhugh** [comment]
Fixed all the comments and added `--max_old_space_size=8192` to the nodejs arguments to prevent it from crashing. Unfortunately there is no great way to reduce the amount of RAM it uses without doing some really funky stuff. Also changed all the hardcoded term from Fall 2017 to Spring 2018. 

Quick note on `dev` vs `prod` mode:

When you run the code in production mode it:
1. Runs a lot faster
2. Totally skips the cache
3. Logs a lot less (Just changed this in the Search NEU codebase and published a new version to npm)

You can run in `prod` mode by setting the env variable `PROD=true` or `NODE_ENV=prod`. It logs which mode it is running in when it first starts up.

**Comment by ryanhugh** [comment]
hahaha I did - just added them back!

**Comment by mnm364** [comment]
Can we disable the cache if `NODE_ENV=development` is that an easy switch you could add, might help with our memory issues...?

**Comment by mnm364** [comment]
You read my mind. Good shit man, thanks!

**Comment by ryanhugh** [comment]
The code behind that is here if you want to take a look!

https://github.com/ryanhugh/searchneu/blob/master/common/abstractMacros.js#L142

**Comment by ryanhugh** [comment]
Hold off on merging for a moment, still debugging stuff haha

**Comment by ryanhugh** [comment]
Ok everything should work! Let me know if it dosen't work on your end. I also pushed some changes to searchneu on npm.

**Comment by noahpresler** [comment]
lgtm

**Comment by noahpresler** [comment]
With confirmation from @mnm364 I will merge and deploy

**Comment by mnm364** [comment]
@ryanhugh did you update the npm package?

**Comment by ryanhugh** [comment]
Yup

**Comment by mnm364** [comment]
lgtm

**Pull request merged**